README.md
# Mangadex@Home

## To Build

### Requirements

- The Java(TM) SE JDK, version 8 or greater
- ~500MB of free space

### Instructions

- Run `./gradlew build` in order to build the entire project
- Find the generated jars in `build/libs`, where the `-all` jar is fat-jar with all dependencies


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [1.1.5] - 2020-07-05
### Added
- [2020-07-05] Added basic graph creation interface to dash [@RedMatriz].

### Changed
- [2020-07-04] Changed GitLab CI to build on every push irregardless of tagging by [@carbotaniuman].
- [2020-07-05] Added `mangadex.network` as allowed domain for `Referer`. Allow blank or missing `Referer` by [@AviKav].
- [2020-07-05] Changed mobile dash to allow for graph creation [@RedMatriz].
- [2020-07-05] Minor improvements to graph load [@RedMatriz].

### Fixed
- [2020-07-04] Fixed typo on `access-control-allow-methods` by [@carbotaniuman].
- [2020-07-05] Fixed minor graph update issues [@RedMatriz].

### Security
- [2020-07-05] Prevent `Referer` matching on subdomains such as `mangadex.org.example.com` by [@AviKav].

## [1.1.4] - 2020-07-04
### Changed
- [2020-07-04] Logging backbone overhauled by [@carbotaniuman].
- [2020-07-04] Bumped client version to `15` by [@carbotaniuman].

### Removed
- [2020-07-04] Rolled back configuration of ServerFilters for CORS by [@carbotaniuman].

## [1.1.3] - 2020-07-04
### Added
- [2020-07-04] Updated to allow for automatic version constant variable by [@carbotaniuman].

### Fixed
- [2020-07-04] Updated Dockerfile to include libsodium for token verification by [@FOG_Yamato].

## [1.1.2] - 2020-07-04
### Fixed
- [2020-07-04] Fixed `client_hostname` default configuration bug by [@carbotaniuman].
- [2020-07-04] Suppressed some non-essential logs by [@carbotaniuman].

## [1.1.1] - 2020-07-04
### Changed
- [2020-07-04] Bumped version number by [@carbotaniuman].

## [1.1.0] - 2020-07-04
### Added
- [2020-06-23] Added Gitlab CI integration by [@lflare].
- [2020-06-28] Added `client_external_port` setting [@wedge1001].
- [2020-06-29] Added rudimentary support of Referer checking to mitigate hotlinking by [@lflare].
- [2020-06-30] Added read and write timeouts to protect against some attacks [@carbotaniuman].
- [2020-06-30] Added `dev_settings` to allow for easier development [@wedge1001].
- [2020-07-02] Added stand-ins for future client control in web interface [@RedMatriz].

### Changed
- [2020-06-28] Added `pasued` field in ServerSettings [@carbotaniuman].
- [2020-06-28] Hopefully fixed connection leaks [@carbotaniuman].
- [2020-07-02] Minor fixes and changes to data handling in web interface [@RedMatriz].
- [2020-07-02] Renamed localstorage keys in web interface [@RedMatriz].
- [2020-06-28] Actually report custom `client_hostname` [@carbotaniuman].

### Fixed
- [2020-06-28] Fixed various state transition bugs by [@carbotaniuman].

### Security
- [2020-07-02] Added option to enforce strict checks on tokens to prevent hotlinking [@carbotaniuman].

## [1.0.0] - 2020-06-22
### Added
- [2020-06-22] Repository clean-up by [@carbotaniuman]
- [2020-06-22] Comments added to JSON configuration file by [@carbotaniuman]
- [2020-06-22] Added UI graph resizing and dragging by [@RedMatriz].
- [2020-06-22] Added client setting generator in UI by [@RedMatriz].

### Changed
- [2020-06-22] Fix incorrect `config.sample.json` naming to `settings.sample.json` by [@lflare]
- [2020-06-22] Logging statements improved by [@carbotaniuman]

## [1.0.0-RC22] - 2020-06-22
### Added
- [2020-06-20] Added `graceful_shutdown_wait_seconds` client setting by [@carbotaniuman]
- [2020-06-20] Added `config.sample.json` sample client configuration file by [@lflare]
- [2020-06-19] Readded ability to set log files level [@carbotaniuman]

## [1.0.0-RC21] - 2020-06-20
### Changed
- [2020-06-20] Fixed logging for DiskLruCache [@carbotaniuman]
- [2020-06-20] Don't automatically delete on corruption [@carbotaniuman]

## [1.0.0-RC20] - 2020-06-19
### Added
- [2020-06-19] Errored out on invalid settings.json tokens [@carbotaniuman]

### Changed
- [2020-06-19] Changed default CPU thread count to `4` by [@lflare].
- [2020-06-19] Removed ability to set log files but increased throughput [@carbotaniuman]

### Fixed
- [2020-06-19] Make graceful shutdown work better [@carbotaniuman]
- [2020-06-19] Actually shutdown logback [@carbotaniuman]

## [1.0.0-RC19] - 2020-06-18
### Added
- [2020-06-16] Added WebUI versions to constants by [@RedMatriz].
- [2020-06-16] Added WebUI PWA support for mobile by [@RedMatriz].
- [2020-06-16] Added WebUI local data caching [@RedMatriz].

### Changed
- [2020-06-16] Reworked graceful shutdown [@carbotaniuman].
- [2020-06-17] Revamped configuration & units by [@lflare].

### Fixed
- [2020-06-17] Fixed minor typo with threads error logging by [@lflare].

## [1.0.0-RC18] - 2020-06-16
### Changed
- [2020-06-16] Changed log level of response timings to INFO by [@lflare].
- [2020-06-16] Added server ping logging [@carbotaniuman].
- [2020-06-16] Added access control headers by [@Fugi].

## [1.0.0-RC17] - 2020-06-15
### Added
- [2020-06-15] Added logging of backend assigned URL to logs by [@lflare].
- [2020-06-15] Added logging of `compromised` softban to logs by [@lflare].

### Changed
- [2020-06-14] Migrated cache metadata over to a sqlite3 handler [@carbotaniuman].
- [2020-06-15] Properly describe dirty builds as dirty by [@lflare].

### Deprecated
- [2020-06-14] Removed old cache subdirectory migration system by [@carbotaniuman].

### Fixed
- [2020-06-14] Switched cache metadata over to a MySql instance [@carbotaniuman].
- [2020-06-15] Fixed tokenized data-saver parser not working by [@lflare].
- [2020-06-15] Properly synchronised sqlite3 handler across threads by [@lflare].

## [1.0.0-RC16] - 2020-06-14
### Added
- [2020-06-14] Added new `client_hostname` selector to allow for custom address binding for Netty by [@lflare].

src/main/kotlin/mdnet/base/Keys.kt
//The code below is adapted from from https://github.com/Mastercard/client-encryption-java/blob/master/src/main/java/com/mastercard/developer/utils/EncryptionUtils.java
//
//Copyright (c) 2019 Mastercard
//
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:
//
//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.
//
//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.
package mdnet.base

import java.io.ByteArrayOutputStream
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.spec.InvalidKeySpecException
import java.security.spec.PKCS8EncodedKeySpec

private const val PKCS_1_PEM_HEADER = "-----BEGIN RSA PRIVATE KEY-----"
private const val PKCS_1_PEM_FOOTER = "-----END RSA PRIVATE KEY-----"
private const val PKCS_8_PEM_HEADER = "-----BEGIN PRIVATE KEY-----"
private const val PKCS_8_PEM_FOOTER = "-----END PRIVATE KEY-----"

fun loadKey(keyDataString: String): PrivateKey? {
    if (keyDataString.contains(PKCS_1_PEM_HEADER)) {
        // OpenSSL / PKCS#1 Base64 PEM encoded file
        val fixedString = keyDataString.replace(PKCS_1_PEM_HEADER, "").replace(PKCS_1_PEM_FOOTER, "")
        return readPkcs1PrivateKey(base64Decode(fixedString))
    }
    if (keyDataString.contains(PKCS_8_PEM_HEADER)) {
        // PKCS#8 Base64 PEM encoded file
        val fixedString = keyDataString.replace(PKCS_8_PEM_HEADER, "").replace(PKCS_8_PEM_FOOTER, "")
        return readPkcs1PrivateKey(base64Decode(fixedString))
    }

    return null
}

private fun readPkcs8PrivateKey(pkcs8Bytes: ByteArray): PrivateKey? {
    val keyFactory = KeyFactory.getInstance("RSA", "SunRsaSign")
    val keySpec = PKCS8EncodedKeySpec(pkcs8Bytes)
    return try {
        keyFactory.generatePrivate(keySpec)
    } catch (e: InvalidKeySpecException) {
        throw IllegalArgumentException("Unexpected key format!", e)
    }
}

private fun readPkcs1PrivateKey(pkcs1Bytes: ByteArray): PrivateKey? {
    // We can't use Java internal APIs to parse ASN.1 structures, so we build a PKCS#8 key Java can understand
    val pkcs1Length = pkcs1Bytes.size
    val totalLength = pkcs1Length + 22
    val pkcs8Header = byteArrayOf(
            0x30, 0x82.toByte(), (totalLength shr 8 and 0xff).toByte(), (totalLength and 0xff).toByte(), // Sequence + total length
            0x2, 0x1, 0x0, // Integer (0)
            0x30, 0xD, 0x6, 0x9, 0x2A, 0x86.toByte(), 0x48, 0x86.toByte(), 0xF7.toByte(), 0xD, 0x1, 0x1, 0x1, 0x5, 0x0, // Sequence: 1.2.840.113549.1.1.1, NULL
            0x4, 0x82.toByte(), (pkcs1Length shr 8 and 0xff).toByte(), (pkcs1Length and 0xff).toByte() // Octet string + length
    )
    val pkcs8bytes = join(pkcs8Header, pkcs1Bytes)
    return readPkcs8PrivateKey(pkcs8bytes)
}

private fun join(byteArray1: ByteArray, byteArray2: ByteArray): ByteArray {
    val bytes = ByteArray(byteArray1.size + byteArray2.size)
    System.arraycopy(byteArray1, 0, bytes, 0, byteArray1.size)
    System.arraycopy(byteArray2, 0, bytes, byteArray1.size, byteArray2.size)
    return bytes
}

private val b64ints = intArrayOf(
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63, 52, 53, 54,
        55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1, -1, 0, 1, 2,
        3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
        20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1, -1, 26, 27, 28, 29, 30,
        31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
        48, 49, 50, 51, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1)

private fun base64Decode(value: String): ByteArray {
    val valueBytes = value.toByteArray()
    val outputStream = ByteArrayOutputStream()
    var i = 0
    while (i < valueBytes.size) {
        var b: Int
        b = if (b64ints[valueBytes[i].toInt()] != -1) {
            b64ints[valueBytes[i].toInt()] and 0xFF shl 18
        } else {
            i++
            continue
        }
        var num = 0
        if (i + 1 < valueBytes.size && b64ints[valueBytes[i + 1].toInt()] != -1) {
            b = b or (b64ints[valueBytes[i + 1].toInt()] and 0xFF shl 12)
            num++
        }
        if (i + 2 < valueBytes.size && b64ints[valueBytes[i + 2].toInt()] != -1) {
            b = b or (b64ints[valueBytes[i + 2].toInt()] and 0xFF shl 6)
            num++
        }
        if (i + 3 < valueBytes.size && b64ints[valueBytes[i + 3].toInt()] != -1) {
            b = b or (b64ints[valueBytes[i + 3].toInt()] and 0xFF)
            num++
        }
        while (num > 0) {
            val c = b and 0xFF0000 shr 16
            outputStream.write(c)
            b = b shl 8
            num--
        }
        i += 4
    }
    return outputStream.toByteArray()
}
